src/Schema/AbstractSchemaManager.php

```diff
@@ -363,9 +363,11 @@ protected function doListTableNames(): array
      * Filters asset names if they are configured to return only a subset of all
      * the found elements.
      *
-     * @param mixed[] $assetNames
+     * @param N $assetNames
```

Comment:
It doesn't look like this method requires a list as input. It can be any collection where the keys will be ignored. It does return a list though.

Reply:
And in practice, it's always called with a list. I don't think there is a way to express the nuance you found while still using the template annotation. Should I switch to `string[]|AbstractAsset[]`? Or should I use `@template N of string|AbstractAsset`, and then `list<N>` and `array`<N>`? Both are technically accurate although less precise.

Reply:
> And in practice, it's always called with a list.

I don't think it matters. The input parameter type should be as permissive as possible, and the return value type should be as specific as possible. This way, the method will be as much usable as possible.

Ideally, it should accept a `string[]` and return a `list<string>`.

> Should I switch to `string[]|AbstractAsset[]`?

It looks like it can filter a collection of any type as long as the predicate accepts this type. We should try to document it this way. Something like:
```php
/**
 * @template T
 * @param T[]
 * @return list<T>
 */
```
At this point, the `$assetNames` parameter name becomes too specific, because it's not necessarily names. We could name it just `$collection`.

Reply:
I don't want to force users to write predicates that will accept `mixed`. Ideally, if they pass a `Predicate<T>`, then the filtering function should accept a `T[]` but I'm not sure if this could be expressed/enforced via the existing type system.

In any event, I wouldn't like to make any code changes just to satisfy the static analysis without understanding the purpose of the API. So far, it doesn't seem like there is such understanding.

Reply:
OK, let me try to understand more then.

Originally, there did not seem to be a way to filter by anything else by the name: https://github.com/doctrine/dbal/pull/98/files#diff-29d1b82efbac329eae1bce15b66d1451dec7dee4a7613c50be9ed426bdad54d6

`filterAssetNames()` returned an array with the same item types as the input. The predicate was internal at that time, hence why having it accept `AbstractAsset` was not such a big deal.

Later on, the possibility was introduced to use user-provided predicates, following https://github.com/doctrine/dbal/issues/3196

That issues makes it clear that the intent was to be able to  build the filter incrementally, giving the possibility to a bundle to add more filters on top of the user-provided configuration.

With that in mind, I now think the implementation in https://github.com/doctrine/dbal/pull/3308 should have restricted the predicate argument to just `string`, and `AbstractSchemaManager::listSequences()` should have been reworked to use `filterAssetNames` with the sequence names, and used the filtered list to return a list of `Sequence` objects. That way, users wouldn't have to handle `AbstractAsset` in their predicate.

I suggest we deprecate passing predicates that support anything else than `string`, and switch to the behavior I'm describing above if they do only support string. This can be figured out with reflection: https://3v4l.org/6VOMH

Reply:
Accepting only strings sounds reasonable. The only downside for that is that if there are tables and sequences with the same names, and we want to allow sequences but ignore tables, it will be impossible to implement.

Another problem with filtering strings is that it is not clear whether we should pass quoted or unquoted names to the filter, the fully qualified names or the shortest possible ones.

It would be much more flexible if the filter accepted assets and filtered them by whatever attributes is needed.

Before we decide on the final approach, I'd like to consider the idea of moving this functionality somewhere else from the DBAL. The only place where the DBAL calls the filter not right before returning the result to the consumer is here: https://github.com/doctrine/dbal/blob/d3aa23fa8916b0deced088ce9581ea2b05e10ef9/src/Schema/AbstractSchemaManager.php#L1224-L1227

So the only use case where the filtering cannot be implemented outside of the DBAL is the schema migration (e.g. when some table exists in the database but it's not managed by the DBAL and thus should be ignored). It's important to ignore the assets at the earliest possible stage of introspection because some of them may be impossible to introspect (see https://github.com/doctrine/dbal/issues/3281#issuecomment-421463994 for reference).

Reply:
:thinking: do you know this "valley of despair" place in the Dunning-Kruger effect curve? I think I've just reached it, because now, I feel like I know too little to take the right decision. I notice also that grep'ing `filter` in the docs directory returns nothing, so let's pause here, and I'll try to document this feature: what is its purpose and current known usages, do we currently pass it quoted or unquoted names (or both :grimacing: ?), why do we sometimes pass objects to it.

> It feels weird to deprive the filter of the information that we already have.

I suppose it is a trade off between giving users full filtering power, and making it easy to write a filter callback.